x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/CommandLineHttpClient.java

```diff
@@ -162,4 +169,84 @@ public String getDefaultURL() {
                 "provide the url", e);
         }
     }
+
+    public static String getErrorCause(HttpResponse httpResponse) {
+        final Object error = httpResponse.getResponseBody().get("error");
+        if (error == null) {
+            return null;
+        }
+        if (error instanceof Map) {
+            Object reason = ((Map) error).get("reason");
+            if (reason != null) {
+                return reason.toString();
+            }
+            final Object root = ((Map) error).get("root_cause");
+            if (root != null && root instanceof Map) {
+                reason = ((Map) root).get("reason");
+                if (reason != null) {
+                    return reason.toString();
+                }
+                final Object type = ((Map) root).get("type");
+                if (type != null) {
+                    return (String) type;
+                }
+            }
+            return String.valueOf(((Map) error).get("type"));
+        }
+        return error.toString();
+    }
+
+    /**
+     * If cluster is not up yet (connection refused or cluster still Red), we will retry @retries number of times
+     */
+    public void checkClusterHealthWithRetriesWaitingForCluster(String username, SecureString password, int retries)
+        throws Exception {
+        final URL clusterHealthUrl = createURL(new URL(getDefaultURL()), "_cluster/health", "?pretty");
+        HttpResponse response = null;
+        try {
+            response = execute("GET", clusterHealthUrl, username, password, () -> null, CommandLineHttpClient::responseBuilder);
+        } catch (Exception e) {
+            if (retries > 0) {
+                Thread.sleep(1000);
+                retries -= 1;
+                checkClusterHealthWithRetriesWaitingForCluster(username, password, retries);
+            } else {
+                throw new IllegalStateException("Failed to determine the health of the cluster. ", e);
+            }
+        }
+        final int responseStatus = response.getHttpStatus();
+        if (responseStatus != HttpURLConnection.HTTP_OK) {
```

Comment:
Technically I would say that 503 is also a condition (in addition to connection problems) that should be retryable.
Cluster health can return it when there's no master (it does some retrying, it's unlikely, but possible).